src/simulation/dynamics/constraintEffector/_UnitTest/test_ConstraintDynamicEffectorUnit.py

```diff
@@ -0,0 +1,331 @@
+    datLog1 = scObject1.scStateOutMsg.recorder()
+    datLog2 = scObject2.scStateOutMsg.recorder()
+    unitTestSim.AddModelToTask(unitTaskName, datLog1)
+    unitTestSim.AddModelToTask(unitTaskName, datLog2)
+
+    # Log energy and momentum variables
+    conservationData1 = scObject1.logger(["totOrbAngMomPntN_N", "totRotAngMomPntC_N", "totOrbEnergy", "totRotEnergy"])
+    conservationData2 = scObject2.logger(["totOrbAngMomPntN_N", "totRotAngMomPntC_N", "totOrbEnergy", "totRotEnergy"])
+    unitTestSim.AddModelToTask(unitTaskName,conservationData1)
+    unitTestSim.AddModelToTask(unitTaskName,conservationData2)
+
+    # Initialize the simulation
+    unitTestSim.InitializeSimulation()
+
+    # Setup and run the simulation
+    stopTime = 1
+    unitTestSim.ConfigureStopTime(macros.sec2nano(stopTime))
+    unitTestSim.ExecuteSimulation()
+
+    # collect the recorded spacecraft states
+    constraintTimeData = datLog1.times() * macros.NANO2SEC
+    r_B1N_N_hist = datLog1.r_BN_N
+    rdot_B1N_N_hist = datLog1.v_BN_N
+    sigma_B1N_hist = datLog1.sigma_BN
+    omega_B1N_B1_hist = datLog1.omega_BN_B
+    r_B2N_N_hist = datLog2.r_BN_N
+    rdot_B2N_N_hist = datLog2.v_BN_N
+    sigma_B2N_hist = datLog2.sigma_BN
+    omega_B2N_B2_hist = datLog2.omega_BN_B
+
+    # collect the logged conservation variables
+    conservationTimeData = conservationData1.times() * macros.NANO2SEC
+    orbEnergy1 = conservationData1.totOrbEnergy
+    orbAngMom1_N = conservationData1.totOrbAngMomPntN_N
+    rotAngMom1PntC1_N = conservationData1.totRotAngMomPntC_N
+    rotEnergy1 = conservationData1.totRotEnergy
+    orbEnergy2 = conservationData2.totOrbEnergy
+    orbAngMom2_N = conservationData2.totOrbAngMomPntN_N
+    rotAngMom2PntC2_N = conservationData2.totRotAngMomPntC_N
+    rotEnergy2 = conservationData2.totRotEnergy
+
+    # Compute constraint violations
+    r_B1N_B1 = np.empty(r_B1N_N_hist.shape)
+    r_B2N_B1 = np.empty(r_B2N_N_hist.shape)
+    r_P2B2_B1 = np.empty(r_B1N_N_hist.shape)
+    for i in range(rotAngMom1PntC1_N.shape[0]):
+        dcm_B1N = RigidBodyKinematics.MRP2C(sigma_B1N_hist[i,:])
+        r_B1N_B1[i,:] = dcm_B1N@r_B1N_N_hist[i,:]
+        r_B2N_B1[i,:] = dcm_B1N@r_B2N_N_hist[i,:]
+        dcm_NB2 = np.transpose(RigidBodyKinematics.MRP2C(sigma_B2N_hist[i,:]))
+        r_P2B2_B1[i,:] = dcm_B1N@dcm_NB2@r_P2B2_B2
+    psi_B1 = r_B1N_B1 + r_P1B1_B1 + r_P2P1_B1Init - (r_B2N_B1 + r_P2B2_B1)
+    sigma_B2B1 = sigma_B2N_hist-sigma_B1N_hist
+
+    # Compute conservation quantities
+    r_CN_N_hist = (r_B1N_N_hist * scObject1.hub.mHub + r_B2N_N_hist * scObject2.hub.mHub)/(scObject1.hub.mHub+scObject2.hub.mHub)
+    r_B1C_N_hist = r_B1N_N_hist - r_CN_N_hist
+    r_B2C_N_hist = r_B2N_N_hist - r_CN_N_hist
+    rdot_CN_N_hist = (rdot_B1N_N_hist * scObject1.hub.mHub + rdot_B2N_N_hist * scObject2.hub.mHub)/(scObject1.hub.mHub+scObject2.hub.mHub)
+    rdot_B1C_N_hist = rdot_B1N_N_hist - rdot_CN_N_hist
+    rdot_B2C_N_hist = rdot_B2N_N_hist - rdot_CN_N_hist
```

Comment:
unused variable, this can be removed?

Reply:
These log computations all feed into the combined conservation quantity calculations. I believe they are required.